clang/lib/3C/ConstraintResolver.cpp

```diff
@@ -373,9 +373,18 @@ CVarSet ConstraintResolver::getExprConstraintVars(Expr *E) {
                        dyn_cast<ArraySubscriptExpr>(UOExpr)) {
           Ret = getExprConstraintVars(ASE->getBase());
         } else {
-          // add a VarAtom to UOExpr's PVConstraint, for &
           CVarSet T = getExprConstraintVars(UOExpr);
           assert("Empty constraint vars in AddrOf!" && !T.empty());
+          // CheckedC prohibits taking the address of a variable with bounds. To
+          // avoid doing this, constrain the target of AddrOf expressions to
+          // PTR. This prevents it from solving to either ARR or NTARR. CheckedC
+          // does permit taking the address of an _Array_ptr when the array
+          // pointer has no declared bounds. With this constraint added however,
+          // 3C will not generate such code.
+          for (auto *CV : T)
+            if (auto *PCV = dyn_cast<PVConstraint>(CV))
+              PCV->constrainOuterTo(CS, CS.getPtr(), true);
```

Comment:
I am wondering: Is this creating the constraint V <: PTR (where V is the constraint variable of the expression) or PTR <: V? I presume it's the latter, but just want to confirm.

Reply:
Yes, PTR <: V. The last parameter for `constrainOuterTo` is a boolean flag `DoLB` that controls this. 
![initial_constraints_graph dot](https://user-images.githubusercontent.com/7480329/104752777-1a34b380-5725-11eb-903d-aa992aabe33b.png)

Reply:
OK, great.

Reply:
Good to merge?

Reply:
Yes.